contracts/RestrictedNORI.sol

```diff
@@ -0,0 +1,1047 @@
+    Removal.ScheduleData memory scheduleData = _removal
+      .getScheduleDataForProjectId(projectId);
+    address recipient = scheduleData.supplierAddress;
+    if (!_scheduleIdToScheduleStruct[projectId].exists) {
+      _createSchedule(projectId);
+    }
+    super._mint(recipient, projectId, amount, "");
+    Schedule storage schedule = _scheduleIdToScheduleStruct[projectId];
+    // slither-disable-next-line unused-return address may already be in set and that is ok
+    schedule.tokenHolders.add(recipient);
+    // slither-disable-next-line unused-return id may already be in the set and that is ok
+    _addressToScheduleIdSet[recipient].add(projectId);
+    return true;
+  }
+
+  /**
+   * Sets up a schedule for the specified project id (implementation).
+   *
+   * @dev schedules are created when removal tokens are listed for sale in the market contract,
+   * so this should only be invoked during `tokensReceived` in the exceptional case that
+   * tokens were sent to this contract without a schedule set up.
+   */
+  function _createSchedule(uint256 projectId) internal {
+    Removal.ScheduleData memory scheduleData = _removal
+      .getScheduleDataForProjectId(projectId);
+    if (scheduleData.startTime == 0) {
+      revert InvalidScheduleStartTime({projectId: projectId});
+    }
+    address recipient = scheduleData.supplierAddress;
+    if (recipient == address(0)) {
+      revert RecipientCannotBeZeroAddress();
+    }
+    if (hasRole(SCHEDULE_CREATOR_ROLE, recipient)) {
+      revert RecipientCannotHaveRole({
+        recipient: recipient,
+        role: "SCHEDULE_CREATOR_ROLE"
+      });
+    }
+    if (_scheduleIdToScheduleStruct[projectId].exists) {
+      revert ScheduleExists({scheduleTokenId: projectId});
+    }
+    Schedule storage schedule = _scheduleIdToScheduleStruct[projectId];
+    // slither-disable-next-line unused-return return value irrelevant
+    _allScheduleIds.add(projectId);
+    RestrictionDuration
+      memory restrictionDuration = getRestrictionDurationForMethodologyAndVersion(
+        scheduleData.methodology,
+        scheduleData.methodologyVersion
+      );
+    if (!restrictionDuration.wasSet) {
+      revert RestrictionDurationNotSet({projectId: projectId});
+    }
+    // slither-disable-next-line unused-return return value irrelevant
+    _addressToScheduleIdSet[recipient].add(projectId);
+    schedule.exists = true;
+    schedule.startTime = scheduleData.startTime;
+    schedule.endTime = scheduleData.startTime + restrictionDuration.duration;
+    schedule.releasedAmountFloor = 0;
+    emit ScheduleCreated(projectId, schedule.startTime, schedule.endTime);
+  }
+
+  /**
+   * @notice Revokes unreleased tokens from a schedule.
+   * This is an *admin* operation callable only by addresses having TOKEN_REVOKER_ROLE
+   * (enforced in `batchRevokeUnreleasedTokenAmounts`)
+   *
+   * @dev Only the maximum revocable number of tokens (unreleased tokens) can be revoked.
```

Comment:
not sure i understand this wording

Reply:
that might be because it's terrible wording. simplified!